python/taichi/lang/kernel_impl.py

```diff
@@ -489,6 +490,17 @@ def func__(*args):
                 if isinstance(needed, template):
                     continue
                 provided = type(v)
+                # Shortened code path for ndarray.
```

Comment:
I'm also curious here. IMHO this modification might make people confused when reading this part of code :-)

Reply:
Yeah that's for sure. I tried to refactor the conditions but got stuck in tests, there were come complex logics that should be simplified. Shall we open another issue to track this?

Reply:
Sounds good.

Reply:
About perf gain, this shortcut will not evaluate `match_ext_arr` and other `isinstance`, so the effect is observable. After all this function body is critical to launch overhead, we should try best to be efficient.

About readability, the first `if` clause in this loop takes a shortcut convention, while the later clauses are nested `if-elif-else`. So I think extract out the Ndarray in a shortcut won't cause confusion. However there's repeated logics later in the clausees, this might be confusing. I'd like to clean this up in anthoer PR.

Reply:
Regarding the logic I think `isinstance(needed, any_arr)` is needed for `Ndarray`s.

Reply:
I see, have amended this issue!